simple dashboard

nikophil opened the pull request:
Hello!

here is a fist PR with a really simple dashboard.

it displays all the configured transports and the number of messages inside them.

it also displays the content of the failure transport, with ability to retry/reject each message.

I've created a simple sandbox app, in which we can easily dispatch messages on amqp, doctrine, redis. It's easier to test the bundle this way, It can sometimes create an error. I can ship it if you're interested.

![Capture du 2019-12-18 12-42-12](https://user-images.githubusercontent.com/10139766/71083771-15a17280-2194-11ea-91f3-de7971df5149.png)

cheers!

nikophil committed:
create first dashboard

nikophil committed:
list failure transport

nikophil committed:
retry message

nikophil committed:
reject message

xosofox commented:
HOLY CRAP!!! Thanks a lot! 🎉 This looks awesome!

I create a even simler https://github.com/karo-io/messenger-demo for testing, so yours sounds/looks way more useful. Is it on github as well?

I'll test your pull request on "mine" and then merge if all works fine

nikophil commented:
i'll share it on github this afternoon.

ho, and a lot of things can/should be improved here, i'll open some discussion in this PR

nikophil commented on DependencyInjection/ReceiverLocatorPass.php:
```diff
@@ -12,16 +12,15 @@ class ReceiverLocatorPass implements CompilerPassInterface
                 $consumeCommandDefinition = $container->getDefinition('console.command.messenger_consume_messages');
                 $names = $consumeCommandDefinition->getArgument(4);
-                $receiverLocatorDefinition->addArgument($names);
+                $receiverLocatorDefinition->replaceArgument(1, $names);
```
can't we use `$container->getExtensionConfig('framework')` and merge all `framework.messenger.transports` we should find in order to remove this hack?

weaverryan replied:
I'm not sure. I'm not sure if I've ever seen that before. In theory, because this is a compiler pass, that should be available (I think?). But we would also be responsible for merging the array of arrays (from multiple config files). I'm not sure it's less of a hack.

nikophil replied:
at least we would not rely on arguments of other services and their order

nikophil replied:
up!

nikophil commented on Locator/FailureTransportLocator.php:
```diff
@@ -0,0 +1,32 @@
+        $this->failureReceiverName = $failureReceiverName;
+    }
+
+    public function getFailureTransport(): TransportInterface
```
i've just realized that it would produce an error if `transport_failure` is null

weaverryan replied:
Let's see if we want to merge this into `ReceiverLocator` and then see if this comment still applies. But yea, if no `failure_transport` is selected, then that whole section of the page should just not show up.

nikophil commented on Locator/FailureTransportLocator.php:
```diff
@@ -0,0 +1,32 @@
+use Symfony\Component\Messenger\Transport\TransportInterface;
+use Symfony\Contracts\Service\ServiceProviderInterface;
+
+class FailureTransportLocator
```
i'm wondering if this should not be merged into `ReceiverLocator`

weaverryan replied:
I'm thinking that too. There is also a PR (that I support in principle) inside Symfony to allow for multiple failure transports - basically on per normal transport. If we want to future-proof a little bit, instead of other classes being able to call `$failureTransportLocator->getFailureTransport()`, we're going to need to keep track of which "failure transport" we're working with and pass that to methods. For example, in `FailedMessageRejecter::rejectFailedMessage()`, we might need a `$failureReceiveName` as an argument. 

If we did that, indeed, we would *only* need the `ReceiverLocator`. Because now code like `FailedMessageRejecter` would be calling something like `$this->receiverLocator->getReceiver($failureReceiveName)` to get the failure transport.

weaverryan commented on Controller/Dashboard.php:
```diff
@@ -0,0 +1,51 @@
+    private $failedMessageRepository;
+
+    public function __construct(
+        Environment $twig,
```
The bundle now depends on `symfony/twig-bundle` - it'll need to be in the composer.json file

weaverryan replied:
And unrelated to your PR, the composer.json file needs a `php` requirement - probably `"php": "^7.1.3",` to match Symfony 4.4

nikophil replied:
should i search for `use Symfony\Component` in the project and add all used components in the `composer.json`?

nikophil replied:
up!

weaverryan replied:
Yea... that would be the *best* thing to do. I usually allow my tests to fail and notify me of any missing dependencies... but it's more responsible to do what you're saying here.

weaverryan commented on Controller/Dashboard.php:
```diff
@@ -0,0 +1,51 @@
+    {
+        $transports = [];
+        foreach ($this->receiverLocator->getReceiversMapping() as $name => $receiver) {
+            $transports[$name] = $receiver instanceof MessageCountAwareInterface ? $receiver->getMessageCount() : -1;
```
`null` might be a better choice than `-1`?

nikophil replied:
sure, i've used the same value than the command `messenger:monitor` without thinking much more thant copy/pasting

weaverryan commented on Controller/Dashboard.php:
```diff
@@ -0,0 +1,51 @@
+
+        try {
+            $failedMessages = $this->failedMessageRepository->listFailedMessages();
+        } catch (\RuntimeException $exception) {
```
Did you mean for this to be a FailureTransportNotListable exception?

weaverryan commented on Exception/FailureTransportNotListable.php:
```diff
@@ -0,0 +1,13 @@
+
+use Symfony\Component\Messenger\Exception\RuntimeException;
+
+class FailureTransportNotListable extends RuntimeException
```
Personal preference (and what's done in Symfony core) - I like the `Exception` suffix - `FailureTransportNotListableException`

weaverryan commented on FailedMessage/FailedMessageRepository.php:
```diff
@@ -0,0 +1,70 @@
+                'class' => \get_class($envelope->getMessage()),
+                'failedAt' => null === $lastRedeliveryStampWithException ? '' : $lastRedeliveryStampWithException->getRedeliveredAt()->format('Y-m-d H:i:s'),
+                'error' => null === $lastRedeliveryStampWithException ? '' : $lastRedeliveryStampWithException->getExceptionMessage(),
+            ];
```
It might be nice to create a `FailedMessageDetails` object here instead of the associate array.

nikophil replied:
:+1:

weaverryan commented on Controller/Dashboard.php:
```diff
@@ -0,0 +1,51 @@
+use Symfony\Component\Messenger\Transport\Receiver\MessageCountAwareInterface;
+use Twig\Environment;
+
+class Dashboard
```
I'd also prefer a `Controller` suffix. I didn't realize at first that a few of these were controllers

weaverryan commented on Resources/views/dashboard.html.twig:
```diff
@@ -0,0 +1,63 @@
+<h1>Messenger monitor dashboard</h1>
+<div class="container">
+
+    {% for label, messages in app.flashes %}
```
You should set a specific flash key and fetch out that specific key here (so we don't accidentally print out other, user-added flashes)

nikophil replied:
sure! nice trick

weaverryan commented on Resources/views/dashboard.html.twig:
```diff
@@ -0,0 +1,63 @@
+                    <td>{{ failedMessage.error }}</td>
+                    <td><a class="btn btn-primary" href="{{ path('karo-io.retry_failed_message', {id: failedMessage.id}) }}">Retry</a></td>
+                    <td><a class="btn btn-danger" href="{{ path('karo-io.reject_failed_message', {id: failedMessage.id}) }}">Reject</a></td>
+                </tr>
```
We can add an `{% else %}` for the "empty" case - like a `<tr><td colspan="6">Your failure transport is empty. You're killing it!</td></tr>`

nikophil replied:
:+1:

nikophil replied:
i didn't know about the for/else syntax!

weaverryan reviewed (commented):
Such a great new chunk of functionality! I would use this already :)

nikophil committed:
get rid of FailureTransportLocator

nikophil committed:
fixed some PR returns

nikophil commented on FailedMessage/FailedMessageRejecter.php:
```diff
@@ -0,0 +1,42 @@
+
+        if (!$failureReceiver instanceof ListableReceiverInterface) {
+            throw new FailureReceiverNotListableException();
+        }
```
This piece of code is repeated three times in the code now.

Any idea to wrap this with an external class?

nikophil replied:
i've created a `FailureReceiverProvider` which also may be usefull in the future when the PR with multiple failure transport will be merged

nikophil commented on FailedMessage/FailedMessageRejecter.php:
```diff
@@ -0,0 +1,42 @@
+    private $receiverLocator;
+    private $failureReceiverName;
+
+    public function __construct(ReceiverLocator $receiverLocator, ?string $failureReceiverName)
```
i'm not really fan of this `?string $failureReceiverName`

nikophil committed:
make internal everything

nikophil committed:
use a value object for the FailureReceiverName

nikophil commented:
@xosofox here is my sandbox

https://github.com/nikophil/messenger-monitor-app

nikophil committed:
add tests

nikophil committed:
Added php-cs-fixer

nikophil committed:
Created FailureReceiverProvider

nikophil commented:
I've added a bunch of tests + phpcs-fixer

I've changed the directory structure to have a `/src` and `/tests` directory. I don't know if that meets the best practices.

I'm not really sure about how to functionally test the controllers, nor how to test if the bundles boots correctly (for example the unit tests pass even if there is a problem in the DI).
Class `FailedMessageRetryer` is also a little bit hard to unit test because we create an actual `Worker` and it needs an event subscriber to stop its work. And after all, testing this class is all about mocking...

We might also create a recipe because the bundle now needs its routing to be loaded.

And finally, at some point, it would be nice to test automatically with travis

cheers!

weaverryan commented on composer.json:
```diff
@@ -3,15 +3,35 @@
+        "symfony/twig-bundle": "^4.4|^5.0"
+    },
+    "require-dev": {
+        "roave/security-advisories": "dev-master",
```
I'm not sure this package makes sense for a library - as packages are more about which is the lowest "allowed" package vs what's installed in the `.lock`file

weaverryan commented on composer.json:
```diff
@@ -3,15 +3,35 @@
+    "autoload-dev": {
+        "psr-4": {
+            "KaroIO\\MessengerMonitorBundle\\": "tests/",
+            "KaroIO\\MessengerMonitorBundle\\Test\\": "src/Test"
```
I would move the one class in this directory somewhere into `tests/` - or even just embed that class at the bottom of any tests that need it. Then we can remove all of this and the `exclude-from-classmap`

weaverryan commented on src/Resources/views/dashboard.html.twig:
```diff
@@ -0,0 +1,69 @@
+</div>
+<script src="https://code.jquery.com/jquery-3.3.1.slim.min.js" integrity="sha384-q8i/X+965DzO0rT7abK41JStQIAqVgRVzpbzo5smXKp4YfRvH+8abtTE1Pi6jizo" crossorigin="anonymous"></script>
+<script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.7/umd/popper.min.js" integrity="sha384-UO2eT0CpHqdSJQ6hJty5KVphtPhzWj9WO1clHTMGa3JDZwrnQq4sF86dIHNDz0W1" crossorigin="anonymous"></script>
+<script src="https://stackpath.bootstrapcdn.com/bootstrap/4.3.1/js/bootstrap.min.js" integrity="sha384-JjSmVgyd0p3pXB1rRibZUAYoIIy6OrQ6VrjIEaFf/nJGzIxFDsf4x0xIM+B07jRM" crossorigin="anonymous"></script>
```
If we don't need these yet, let's not include them... yet

weaverryan commented on src/Exception/ReceiverDoesNotExistException.php:
```diff
@@ -0,0 +1,23 @@
+/**
+ * @internal
+ */
+class ReceiverDoesNotExistException extends RuntimeException
```
We should consider adding `final` to any classes that we can. And if there are a few that we would like to make final, but cannot because we're mocking them, add an `@final` annotation instead.

nikophil replied:
haha i HAVE phpstorm configured to add it automatically, but IIRC few years ago you told me that you don't do that in symfony (or you USED to not to that)
so, i methodically removed them! :joy:

weaverryan replied:
whoooops :D 

We may not have then... who knows. We're not "absolute" about it. But because it "minimizes" the API (especially when applied to non-internal classes), I like to add it.

weaverryan commented on src/FailureReceiver/FailureReceiverName.php:
```diff
@@ -0,0 +1,23 @@
+{
+    private $failureReceiverName;
+
+    public function __construct(?string $failureReceiverName = null)
```
I don't think the arg needs to be optional. Even if there is no failure receiver, we still pass null to this I believe in the compiler pass.

weaverryan commented on src/Locator/ReceiverLocator.php:
```diff
@@ -0,0 +1,48 @@
+    private $receiverLocator;
+    private $receiverNames;
+
+    public function __construct(ServiceProviderInterface $receiverLocator, $receiverNames = [])
```
Does `$receiverNames` need to be optional? And should it have an `array` type hint?

weaverryan commented on src/Resources/config/services.xml:
```diff
@@ -0,0 +1,69 @@
+        </service>
+
+        <service id="karo-io.controller.reject_failed_message" class="KaroIO\MessengerMonitorBundle\Controller\RejectFailedMessageController">
+            <tag name="controller.service_arguments" />
```
I don't think we need this tag on any of the controllers because you're using proper DI (as you should, because this is a 3rd party lib and there is only 1 action per controller) and not relying on action injection.

weaverryan commented on src/FailureReceiver/FailureReceiverProvider.php:
```diff
@@ -0,0 +1,37 @@
+use KaroIO\MessengerMonitorBundle\Locator\ReceiverLocator;
+use Symfony\Component\Messenger\Transport\Receiver\ListableReceiverInterface;
+
+class FailureReceiverProvider
```
Make this `@internal`

nikophil replied:
`@internal` + `@final` ? it's mocked somewhere

weaverryan replied:
Yep, both are fine - they mean subtly different things. The `@final` will actually prevent *us* from accidentally sub-classing it in the bundle :)

weaverryan reviewed (commented):
Minor stuff - great work!

nikophil committed:
fixes after PR

weaverryan commented on composer.json:
```diff
@@ -3,15 +3,39 @@
+        },
+        "exclude-from-classmap": [
+            "src/Test/"
+        ]
```
Not needed anymore

weaverryan reviewed (approved):
Very nice - this is ready to go

weaverryan commented:
Hey @xosofox!

Are you available these days to merge this? Or... allow us access to the repo 😇 ?

Thanks!

xosofox commented:
Hey guys, sorry I fell off the grid for some time...

This is huge - thanks a lot for the effort!! 🎉

weaverryan commented:
I understand completely :). @nikophil has continued his work in a big way over on his fork. How would you like to proceed? Should we keep that fork? Or make a PR back here when things are ready? We’ve been working piece-by-piece over there.

Thanks!

xosofox commented:
I'm currently lacking a project to test it/work on it/use it - so my "urge" to help out is too small, unfortunately, to keep up with the pace.

Short: continue over there, no problem, it might not make sense over here anymore 🤷‍♂

weaverryan commented:
Thanks @xosofox - and thanks for bootstrapping this ❤️

xosofox commented:
It's my pleasure
I added a link on the master README so people have a chance to find @nikophil 's work
